upgrading/upgrade-runnning-applications/4_migrate_contract.md

```diff
@@ -0,0 +1,138 @@
+In the previous step you made the two versions of the `create-daml-app` and `forum` packages
+available as imports.  Now we can write the contract to migrate between the two packages.
+
+In the IDE, open a new file `../migration/daml/Migrate.daml`{{open}} and paste the following code.
+Read it carefully, the comments will help you understand how this module is working.
+
+<pre class="file" data-filename="../migration/daml/Migrate.daml" data-target="append">
+
+module Migrate where
+
+-- Notice the renamed module names
+import qualified V1.User
+import qualified V2.User
+import qualified V1.Forum
+import qualified V2.Forum
+
+-- These are pure functions describing how data of V1 is moved to V2 and back.
+upgradeUser : V1.User.User -> V2.User.User
+upgradeUser V1.User.User{..} = V2.User.User with nickname = None, ..
+
+rollbackUser : V2.User.User -> V1.User.User
+rollbackUser V2.User.User{..} = V1.User.User with ..
+
+upgradePost : V1.Forum.Post -> V2.Forum.Post
+upgradePost V1.Forum.Post{..} = V2.Forum.Post with user = upgradeUser user, ..
+
+rollbackPost : V2.Forum.Post -> V1.Forum.Post
+rollbackPost V2.Forum.Post{..} = V1.Forum.Post with user = rollbackUser user, ..
+
+upgradeComment : V1.Forum.Comment -> ContractId V2.Forum.Post -> V2.Forum.Comment
+upgradeComment V1.Forum.Comment{..} newPostCid
+  = V2.Forum.Comment with commenter = upgradeUser commenter, post = newPostCid, ..
+
+rollbackComment : V2.Forum.Comment -> ContractId V1.Forum.Post -> V1.Forum.Comment
+rollbackComment V2.Forum.Comment{..} oldPostCid
+  = V1.Forum.Comment with commenter = rollbackUser commenter, post = oldPostCid, ..
+
```

Comment:
Can you explain here why upgrade and rollback are needed and what each does? I know it's trivial, but we have to spell it out to the user.

Reply:
It's explained in the last bullet point on that page, would you like to move it further up?